src/crimson/osd/osd_operation.cc

```diff
@@ -50,44 +50,60 @@ void OSDOperationRegistry::put_historic(const ClientRequest& op)
 {
   using crimson::common::local_conf;
   // unlink the op from the client request registry. this is a part of
-  // the re-link procedure. finally it will be in historic registry.
-  constexpr auto client_reg_index =
-    static_cast<size_t>(OperationTypeCode::client_request);
+  // the re-link procedure. finally it will be in historic/historic_slow registry.
   constexpr auto historic_reg_index =
     static_cast<size_t>(OperationTypeCode::historic_client_request);
   constexpr auto slow_historic_reg_index = 
     static_cast<size_t>(OperationTypeCode::historic_slow_client_request);
 
-  auto& client_registry = get_registry<client_reg_index>();
-  // slow ops should put into historic_slow_client_request list
-  // we only save the newest op
   if (get_duration(op) > local_conf()->osd_op_complaint_time) {
     auto& slow_historic_registry = get_registry<slow_historic_reg_index>();
-    slow_historic_registry.splice(std::end(slow_historic_registry),
-          client_registry,
-          client_registry.iterator_to(op));
-
-    if (num_slow_ops >= local_conf()->osd_op_history_slow_op_size) {
-      slow_historic_registry.pop_front();
-    } else {
-      ++num_slow_ops;
-    }
+    _put_historic(slow_historic_registry,
+      op,
+      local_conf()->osd_op_history_slow_op_size,
+      true);
   } else {
     auto& historic_registry = get_registry<historic_reg_index>();
-    historic_registry.splice(std::end(historic_registry),
-          client_registry,
-          client_registry.iterator_to(op));
+    _put_historic(historic_registry,
+      op,
+      local_conf()->osd_op_history_size,
+      false);
+  }
+}
 
-    if (num_recent_ops >= local_conf()->osd_op_history_size) {
-      historic_registry.pop_front();
+void OSDOperationRegistry::_put_historic(
+  op_list& list,
+  const class ClientRequest& op,
+  uint64_t max, bool slow)
+{
+  constexpr auto client_reg_index =
+    static_cast<size_t>(OperationTypeCode::client_request);
+  auto& client_registry = get_registry<client_reg_index>();
+
+  // slow ops should put into historic_slow_client_request list
+  // we only save the newest op
+  list.splice(std::end(list), client_registry, client_registry.iterator_to(op));
+  ClientRequest::ICRef(
+      &op, /* add_ref= */true
+    ).detach(); // yes, "leak" it for now!
+
+  bool need_trim = (num_slow_ops >= max && slow) ||
+    (num_recent_ops >= max && !slow)
+    ? true : false;
+  if (need_trim) {
+    auto old_op_ptr = &list.front();
+    list.pop_front();
+    const auto& old_op =
+      static_cast<const ClientRequest&>(*old_op_ptr);
+    // clear a previously "leaked" op
+    ClientRequest::ICRef(&old_op, /* add_ref= */false);
+  } else {
```

Comment:
The `if (need_trim)` case should appear unconditionally since we already updated `list` earlier:

```
    if (slow) {
      ++num_slow_ops;
    } else {
      ++num_recent_ops;
    }

    bool need_trim = ...
    if (need_trim) {
      ..
    }

Reply:
ah, ok, my logic is correct, but your logic will be easier to understand

Reply:
if `needs_trim` is true, the total number of ops (slow and recent) is the same - that's true.
However, we need to differentiate between the two to keep the counting. Assuming:

num_slow_ops = n
num_recent_ops = m
needs_trim = true

We trim the oldest op (front) and push back the newest. There are three possible scenarios now:

1. slow_op was trimmed and recent op was added:
    num_slow_ops = n - 1
    num_recent_ops = m + 1

2. recent_op was trimmed and recent op was added:
    num_slow_ops = n + 1
    num_recent_ops = m - 1

3. the same op that was trimmed is the same as the one that was added:
    num_slow_ops = n
    num_recent_ops = m

Reply:
Can we use the list size to avoid the counters?

Reply:
> Can we use the list size to avoid the counters?

ah yes, use list size would be better